In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn import ensemble, linear_model
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import mean_squared_error

import warnings
warnings.filterwarnings('ignore')

pd.set_option('display.max_columns', None)
import plotly.graph_objs as go
import plotly.offline as py


pd.set_option('display.max_columns', None)  
pd.set_option('display.max_rows', 200)  

## Import the data

In [2]:
train = pd.read_csv('../Bain women datahack - QL/train.csv')
test = pd.read_csv('../Bain women datahack - QL/test.csv')

ss = pd.read_csv('../Bain women datahack - QL/sample_submission.csv')

In [3]:
train.head()

ID  Day_No  Course_ID Course_Domain Course_Type  Short_Promotion  \
0   1       1          1   Development      Course                0   
1   2       2          1   Development      Course                0   
2   3       3          1   Development      Course                0   
3   4       4          1   Development      Course                0   
4   5       5          1   Development      Course                0   

   Public_Holiday  Long_Promotion  User_Traffic  Competition_Metric  Sales  
0               1               1         11004               0.007     81  
1               0               1         13650               0.007     79  
2               0               1         11655               0.007     75  
3               0               1         12054               0.007     80  
4               0               1          6804               0.007     41

In [4]:
train.Course_Domain.unique()

array(['Development', 'Software Marketing', 'Finance & Accounting',
       'Business'], dtype=object)

In [5]:
train.Course_Type.unique()

array(['Course', 'Program', 'Degree'], dtype=object)

The percentage of missing values

In [6]:
print(train.isnull().sum()/train.shape[0])
print()
print(test.isnull().sum()/test.shape[0])

ID                    0.000000
Day_No                0.000000
Course_ID             0.000000
Course_Domain         0.000000
Course_Type           0.000000
Short_Promotion       0.000000
Public_Holiday        0.000000
Long_Promotion        0.000000
User_Traffic          0.000000
Competition_Metric    0.003445
Sales                 0.000000
dtype: float64

ID                    0.000000
Day_No                0.000000
Course_ID             0.000000
Course_Domain         0.000000
Course_Type           0.000000
Short_Promotion       0.000000
Public_Holiday        0.000000
Long_Promotion        0.000000
Competition_Metric    0.003333
dtype: float64


Impute the missing values with the mean value

In [7]:
train['Competition_Metric'].fillna(train['Competition_Metric'].mean(), inplace = True)
test['Competition_Metric'].fillna(test['Competition_Metric'].mean(), inplace = True)

Impute the test **`User_Traffic`** with train data

In [8]:
CourseID_Traffic_DF = pd.DataFrame(train.groupby(['Course_ID'])['User_Traffic'].mean())

In [9]:
test = test.join(CourseID_Traffic_DF,on = 'Course_ID', how = 'left')

In [10]:
test.head()

ID  Day_No  Course_ID Course_Domain Course_Type  Short_Promotion  \
0  883     883          1   Development      Course                1   
1  884     884          1   Development      Course                1   
2  885     885          1   Development      Course                1   
3  886     886          1   Development      Course                1   
4  887     887          1   Development      Course                0   

   Public_Holiday  Long_Promotion  Competition_Metric  User_Traffic  
0               0               1               0.007  11882.071429  
1               0               1               0.007  11882.071429  
2               0               1               0.007  11882.071429  
3               0               1               0.007  11882.071429  
4               0               1               0.007  11882.071429

### Get the target y and combine training and testing


In [11]:
y_train = np.log1p(train['Sales'])
# train.drop(['Sales'], inplace = True, axis = 1)

test['Sales'] = 0

data = pd.concat([train,test])


In [12]:
data.dtypes

Competition_Metric    float64
Course_Domain          object
Course_ID               int64
Course_Type            object
Day_No                  int64
ID                      int64
Long_Promotion          int64
Public_Holiday          int64
Sales                   int64
Short_Promotion         int64
User_Traffic          float64
dtype: object

In [13]:
data.head(3)

Competition_Metric Course_Domain  Course_ID Course_Type  Day_No  ID  \
0               0.007   Development          1      Course       1   1   
1               0.007   Development          1      Course       2   2   
2               0.007   Development          1      Course       3   3   

   Long_Promotion  Public_Holiday  Sales  Short_Promotion  User_Traffic  
0               1               1     81                0       11004.0  
1               1               0     79                0       13650.0  
2               1               0     75                0       11655.0

In [14]:
test.Day_No.describe()

count    36000.000000
mean       912.500000
std         17.318343
min        883.000000
25%        897.750000
50%        912.500000
75%        927.250000
max        942.000000
Name: Day_No, dtype: float64

The test date ranges from day #883 to #942, for 60 days

# Feature engineering

Change the **`Course_Domain`** and **`Course_Type`** to the overall average sales. This is to change the categories to data.

In [15]:
Course_Domain_sales_mean = pd.DataFrame(data.groupby(['Course_Domain'])['Sales'].mean())
Course_Domain_sales_mean.rename(columns = {'Sales':'Course_Domain_sales_mean'}, inplace = True)

Course_Type_sales_mean = pd.DataFrame(data.groupby(['Course_Type'])['Sales'].mean())
Course_Type_sales_mean.rename(columns = {'Sales':'Course_Type_sales_mean'}, inplace = True)


In [16]:
data = data.join(Course_Domain_sales_mean, on = 'Course_Domain',how = 'left')
data = data.join(Course_Type_sales_mean, on = 'Course_Type',how = 'left')

In [17]:
# remove the original category
data.drop(['Course_Domain','Course_Type'], axis = 1, inplace = True)

In [18]:
# averaged sales with groupby course ID and public holiday
data['sales_meanT7_holiday'] = data.groupby(['Course_ID', 'Public_Holiday'])['Sales'].transform(lambda x: x.shift(60).rolling(7).mean())
data['sales_stdT7_holiday'] = data.groupby(['Course_ID', 'Public_Holiday'])['Sales'].transform(lambda x: x.shift(60).rolling(7).std())

data['sales_meanT30_holiday'] = data.groupby(['Course_ID', 'Public_Holiday'])['Sales'].transform(lambda x: x.shift(60).rolling(30).mean())
data['sales_stdT30_holiday'] = data.groupby(['Course_ID', 'Public_Holiday'])['Sales'].transform(lambda x: x.shift(60).rolling(30).std())

data['sales_meanT60_holiday'] = data.groupby(['Course_ID', 'Public_Holiday'])['Sales'].transform(lambda x: x.shift(60).rolling(60).mean())
data['sales_stdT60_holiday'] = data.groupby(['Course_ID', 'Public_Holiday'])['Sales'].transform(lambda x: x.shift(60).rolling(60).std())


In [19]:
# average sales with rolling data
data['sales_rolling_mean_t7'] = data.groupby(['Course_ID'])['Sales'].transform(lambda x: x.shift(60).rolling(7).mean())
data['sales_rolling_std_t7'] = data.groupby(['Course_ID'])['Sales'].transform(lambda x: x.shift(60).rolling(7).std())
data['sales_rolling_mean_t30'] = data.groupby(['Course_ID'])['Sales'].transform(lambda x: x.shift(60).rolling(30).mean())
data['sales_rolling_std_t30'] = data.groupby(['Course_ID'])['Sales'].transform(lambda x: x.shift(60).rolling(30).std())
data['sales_rolling_skew_t30'] = data.groupby(['Course_ID'])['Sales'].transform(lambda x: x.shift(60).rolling(30).skew())
data['sales_rolling_kurt_t30'] = data.groupby(['Course_ID'])['Sales'].transform(lambda x: x.shift(60).rolling(30).kurt())

data['sales_rolling_mean_t60'] = data.groupby(['Course_ID'])['Sales'].transform(lambda x: x.shift(60).rolling(60).mean())
data['sales_rolling_std_t60'] = data.groupby(['Course_ID'])['Sales'].transform(lambda x: x.shift(60).rolling(60).std())
data['sales_rolling_skew_t60'] = data.groupby(['Course_ID'])['Sales'].transform(lambda x: x.shift(60).rolling(60).skew())
data['sales_rolling_kurt_t60'] = data.groupby(['Course_ID'])['Sales'].transform(lambda x: x.shift(60).rolling(60).kurt())

From the **`Day_No`** we can extract the **`date_in_year`**, **`date_in_month`**, and **`date_in_week`**, assuming there is a pattern with 365, 30, and 7 days cycle

In [20]:
# date info
data['date_in_year'] = data.Day_No%365
data['date_in_month'] = data.Day_No%30
data['date_in_week'] = data.Day_No%7

# remove original day_no info
data.drop(['Day_No'], axis = 1, inplace = True)


In [21]:
data.head(100)

Competition_Metric  Course_ID   ID  Long_Promotion  Public_Holiday  Sales  \
0                0.007          1    1               1               1     81   
1                0.007          1    2               1               0     79   
2                0.007          1    3               1               0     75   
3                0.007          1    4               1               0     80   
4                0.007          1    5               1               0     41   
5                0.007          1    6               1               0     62   
6                0.007          1    7               1               0    122   
7                0.007          1    8               1               0    114   
8                0.007          1    9               1               0    121   
9                0.007          1   10               1               0    100   
10               0.007          1   11               1               0     86   
11               0.007          1   12               1               0     47   
12               0.007          1   13               1               0     85   
13               0.007          1   14               1               0     72   
14               0.007          1   15               1               0     74   
15               0.007          1   16               1               0     76   
16               0.007          1   17               1               0     53   
17               0.007          1   18               1               0     61   
18               0.007          1   19               1               0     42   
19               0.007          1   20               1               0     51   
20               0.007          1   21               1               0    114   
21               0.007          1   22               1               0    104   
22               0.007          1   23               1               0    107   
23               0.007          1   24               1               0     92   
24               0.007          1   25               1               0     82   
25               0.007          1   26               1               0     39   
26               0.007          1   27               1               0     80   
27               0.007          1   28               1               0     63   
28               0.007          1   29               1               0     61   
29               0.007          1   30               1               0     80   
30               0.007          1   31               1               0     78   
31               0.007          1   32               1               0     81   
32               0.007          1   33               1               0     53   
33               0.007          1   34               1               0     57   
34               0.007          1   35               1               0    117   
35               0.007          1   36               1               0    113   
36               0.007          1   37               1               0    116   
37               0.007          1   38               1               0    100   
38               0.007          1   39               1               0     76   
39               0.007          1   40               1               0     55   
40               0.007          1   41               1               0     77   
41               0.007          1   42               1               0     77   
42               0.007          1   43               1               0     71   
43               0.007          1   44               1               0     90   
44               0.007          1   45               1               0     80   
45               0.007          1   46               1               0     68   
46               0.007          1   47               1               0     42   
47               0.007          1   48               1               0     64   
48               0.007          1  

In [22]:
data.dtypes

Competition_Metric          float64
Course_ID                     int64
ID                            int64
Long_Promotion                int64
Public_Holiday                int64
Sales                         int64
Short_Promotion               int64
User_Traffic                float64
Course_Domain_sales_mean    float64
Course_Type_sales_mean      float64
sales_meanT7_holiday        float64
sales_stdT7_holiday         float64
sales_meanT30_holiday       float64
sales_stdT30_holiday        float64
sales_meanT60_holiday       float64
sales_stdT60_holiday        float64
sales_rolling_mean_t7       float64
sales_rolling_std_t7        float64
sales_rolling_mean_t30      float64
sales_rolling_std_t30       float64
sales_rolling_skew_t30      float64
sales_rolling_kurt_t30      float64
sales_rolling_mean_t60      float64
sales_rolling_std_t60       float64
sales_rolling_skew_t60      float64
sales_rolling_kurt_t60      float64
date_in_year                  int64
date_in_month               

In [23]:
# User_Traffic is dropped because it has too much correlation with Sales. Therefore it is considered as a "y" value
drop_list = ['ID','Course_ID','Sales','User_Traffic']
data.drop(drop_list, axis = 1, inplace = True)

In [24]:
train_data = data[:train.shape[0]]
test_data = data[train.shape[0]:]


# Modeling

## LightGBM with Cross validation

In [25]:
categoricals = ['Short_Promotion','Public_Holiday','Long_Promotion','date_in_year','date_in_month','date_in_week']

## Tune hyperparameters

In [26]:
from sklearn.model_selection import GridSearchCV
import lightgbm as lgb

model_lgb = lgb.LGBMRegressor(boosting_type = 'gbdt', learning_rate = 0.05, num_leaves = 136, n_estimators = 90)

params_test1 = {
    'num_leaves': range(130,140,2)
}

params_test2 = {
    'n_estimators': range(30,100,10)
}

params_test3 = {
    'subsample': np.arange(0.5,1,0.1),
    'subsample_freq': range(2,10,2)
}


In [27]:
# gsearch1 = GridSearchCV(estimator = model_lgb, param_grid = params_test3, 
#                        scoring = 'neg_mean_squared_error', cv = 3)

# gsearch1.fit(train_data, y_train)
# print(gsearch1.best_estimator_)
# print(gsearch1.best_score_)

In [28]:
params = {
    'boosting_type': 'gbdt',
    'objective': 'poisson',
    'metric': 'rmse',
    'subsample_freq': 5,
    'learning_rate': 0.05,
    'num_leaves': 2**8,
    'feature_fraction': 0.9,
    'bagging_fraction':0.9,
    'reg_lambda':2,
    'reg_alpha':1
}

In [29]:
from sklearn.model_selection import KFold

kf = KFold ( n_splits = 5, shuffle = True, random_state = 40)


In [30]:
train.shape

(512087, 11)

In [31]:
import lightgbm as lgb

In [32]:
models_lgb = []
for train_index, test_index in kf.split(train_data): # train_index 就是train的index
    train_features = train_data.loc[train_index]
    train_y = y_train.loc[train_index]
    
    test_features = train_data.loc[test_index]
    test_y = y_train.loc[test_index]
    
    d_training = lgb.Dataset(train_features, label = train_y, 
                             categorical_feature = categoricals)
    d_test = lgb.Dataset(test_features, label = test_y, 
                            categorical_feature = categoricals)
    
    model = lgb.train(params, train_set = d_training, num_boost_round = 20000,
                     valid_sets = [d_training, d_test], verbose_eval = 100, 
                     early_stopping_rounds = 50)
    models_lgb.append(model)
#     del train_features, train_y, test_features, test_y, d_training, d_test
    

Training until validation scores don't improve for 50 rounds
[100]	training's rmse: 0.176402	valid_1's rmse: 0.182036
[200]	training's rmse: 0.151028	valid_1's rmse: 0.159898
[300]	training's rmse: 0.138797	valid_1's rmse: 0.149999
[400]	training's rmse: 0.132978	valid_1's rmse: 0.146069
[500]	training's rmse: 0.129557	valid_1's rmse: 0.144285
[600]	training's rmse: 0.126265	valid_1's rmse: 0.142557
[700]	training's rmse: 0.123702	valid_1's rmse: 0.141337
[800]	training's rmse: 0.121647	valid_1's rmse: 0.140497
[900]	training's rmse: 0.119983	valid_1's rmse: 0.13999
[1000]	training's rmse: 0.118394	valid_1's rmse: 0.139539
[1100]	training's rmse: 0.116787	valid_1's rmse: 0.138957
[1200]	training's rmse: 0.115256	valid_1's rmse: 0.138462
[1300]	training's rmse: 0.114121	valid_1's rmse: 0.138266
[1400]	training's rmse: 0.112722	valid_1's rmse: 0.137827
[1500]	training's rmse: 0.111509	valid_1's rmse: 0.137484
[1600]	training's rmse: 0.110298	valid_1's rmse: 0.137147
[1700]	training's rms

Early stopping, best iteration is:
[3744]	training's rmse: 0.0938058	valid_1's rmse: 0.133032
Training until validation scores don't improve for 50 rounds
[100]	training's rmse: 0.174644	valid_1's rmse: 0.178632
[200]	training's rmse: 0.149538	valid_1's rmse: 0.156776
[300]	training's rmse: 0.136929	valid_1's rmse: 0.146569
[400]	training's rmse: 0.131463	valid_1's rmse: 0.142977
[500]	training's rmse: 0.128246	valid_1's rmse: 0.141355
[600]	training's rmse: 0.12512	valid_1's rmse: 0.13968
[700]	training's rmse: 0.122765	valid_1's rmse: 0.138602
[800]	training's rmse: 0.120725	valid_1's rmse: 0.137746
[900]	training's rmse: 0.119225	valid_1's rmse: 0.137316
[1000]	training's rmse: 0.117645	valid_1's rmse: 0.136828
[1100]	training's rmse: 0.1159	valid_1's rmse: 0.136122
[1200]	training's rmse: 0.114381	valid_1's rmse: 0.13559
[1300]	training's rmse: 0.113103	valid_1's rmse: 0.135231
[1400]	training's rmse: 0.11188	valid_1's rmse: 0.134903
[1500]	training's rmse: 0.110646	valid_1's rmse:

In [33]:
results_lgb = []
for model in models_lgb:
    if  results_lgb == []:
        results_lgb = np.expm1(model.predict(test_data, num_iteration=model.best_iteration)) / len(models_lgb)
    else:
        results_lgb += np.expm1(model.predict(test_data, num_iteration=model.best_iteration)) / len(models_lgb)
    del model

In [34]:
results_lgb

array([141.67279403, 120.0289671 , 108.49097282, ..., 158.38137998,
       163.82001975, 184.59993854])

In [35]:
ss.Sales = results_lgb
ss.to_csv("submission_lgb_Yeqing_Xia.csv", index=False)

## XGBoost

In [36]:
import xgboost as xgb
from xgboost import XGBRegressor

params = {'max_depth': 10, 
          'learning_rate': 0.1, 
          'objective': 'reg:squarederror', 
          'subsample': 0.85, 
          'colsample_bytree': 0.85,
          'n_estimators': 4000,
         'eval_metric': 'rmse'
         }


In [37]:
models_xgb = []
for train_index, test_index in kf.split(train_data): # train_index 就是train的index
    train_features = train_data.loc[train_index]
    train_y = y_train.loc[train_index]
    
    test_features = train_data.loc[test_index]
    test_y = y_train.loc[test_index]
    
    d_training = xgb.DMatrix(train_features, label = train_y)
    d_test = xgb.DMatrix(test_features, label = test_y)
    
    evallist = [(d_training, 'train'), (d_test, 'eval')]  # 这里顺序变了，会导致下面early stopping的metric变化，按照后一个做metric
    
    model = xgb.train(params, d_training, num_boost_round = 10000, evals = evallist, 
                      verbose_eval = 100, early_stopping_rounds = 100)
    
    models_xgb.append(model)


[0]	train-rmse:3.80846	eval-rmse:3.80996
Multiple eval metrics have been passed: 'eval-rmse' will be used for early stopping.

Will train until eval-rmse hasn't improved in 100 rounds.
[100]	train-rmse:0.13463	eval-rmse:0.148556
[200]	train-rmse:0.116193	eval-rmse:0.136959
[300]	train-rmse:0.107018	eval-rmse:0.132803
[400]	train-rmse:0.100356	eval-rmse:0.130482
[500]	train-rmse:0.094839	eval-rmse:0.129106
[600]	train-rmse:0.090203	eval-rmse:0.128125
[700]	train-rmse:0.086337	eval-rmse:0.127598
[800]	train-rmse:0.083406	eval-rmse:0.127629
Stopping. Best iteration:
[703]	train-rmse:0.086213	eval-rmse:0.12756

[0]	train-rmse:3.80882	eval-rmse:3.80866
Multiple eval metrics have been passed: 'eval-rmse' will be used for early stopping.

Will train until eval-rmse hasn't improved in 100 rounds.
[100]	train-rmse:0.13308	eval-rmse:0.146661
[200]	train-rmse:0.116329	eval-rmse:0.136478
[300]	train-rmse:0.106671	eval-rmse:0.131959
[400]	train-rmse:0.10004	eval-rmse:0.129519
[500]	train-rmse:0.094

In [38]:
results_xgb = []
for model in models_xgb:
    if  results_xgb == []:
        results_xgb = np.expm1(model.predict(xgb.DMatrix(test_data))) / len(models_xgb)
    else:
        results_xgb += np.expm1(model.predict(xgb.DMatrix(test_data))) / len(models_xgb)
    del model

In [39]:
results_xgb

array([115.71158, 113.24753, 105.97563, ..., 143.9178 , 150.9237 ,
       163.62325], dtype=float32)

In [40]:
ss.Sales = results_xgb
ss.to_csv("submission_xgb_Yeqing_Xia.csv", index=False)

## CatBoost

In [41]:
from catboost import Pool, CatBoostRegressor


In [47]:
models_cat = []
for train_index, test_index in kf.split(train_data): # train_index 就是train的index
    train_features = train_data.loc[train_index]
    train_y = y_train.loc[train_index]
    
    test_features = train_data.loc[test_index]
    test_y = y_train.loc[test_index]
    
    
    d_training = Pool(train_features, label = train_y)
    d_test = Pool(test_features, label = test_y)
    
    model = CatBoostRegressor(loss_function="RMSE",
                           eval_metric="RMSE",
#                            task_type="GPU",
                           learning_rate=0.05,
                           iterations=1500,
                           l2_leaf_reg=5,
                           random_seed=42,
                           od_type="Iter",
                           depth=10,
                           early_stopping_rounds=1000,
                           border_count=32,
                           subsample = 0.85,
                          )
    model.fit(d_training,
                    eval_set=d_test,
                    use_best_model=True,
                    verbose=100)
    models_cat.append(model)    

0:	learn: 0.4474157	test: 0.4476696	best: 0.4476696 (0)	total: 64.9ms	remaining: 1m 37s
100:	learn: 0.2011888	test: 0.2020773	best: 0.2020773 (100)	total: 6.13s	remaining: 1m 24s
200:	learn: 0.1798476	test: 0.1815364	best: 0.1815364 (200)	total: 12.1s	remaining: 1m 18s
300:	learn: 0.1691168	test: 0.1715177	best: 0.1715177 (300)	total: 18.1s	remaining: 1m 11s
400:	learn: 0.1624693	test: 0.1654908	best: 0.1654908 (400)	total: 24.2s	remaining: 1m 6s
500:	learn: 0.1574480	test: 0.1610297	best: 0.1610297 (500)	total: 30.7s	remaining: 1m 1s
600:	learn: 0.1537256	test: 0.1578380	best: 0.1578380 (600)	total: 37.9s	remaining: 56.6s
700:	learn: 0.1506755	test: 0.1553689	best: 0.1553689 (700)	total: 44.9s	remaining: 51.2s
800:	learn: 0.1481696	test: 0.1534132	best: 0.1534132 (800)	total: 51.5s	remaining: 45s
900:	learn: 0.1459441	test: 0.1516998	best: 0.1516998 (900)	total: 58.4s	remaining: 38.8s
1000:	learn: 0.1438532	test: 0.1500876	best: 0.1500876 (1000)	total: 1m 5s	remaining: 32.4s
1100:	lea

In [48]:
results_cat = []
for model in models_cat:
    if  results_cat == []:
        results_cat = np.expm1(model.predict(test_data)) / len(models_cat)
    else:
        results_cat += np.expm1(model.predict(test_data)) / len(models_cat)
    del model

In [49]:
results_cat

array([122.25848491, 108.23339846, 103.06185584, ..., 144.05492245,
       141.66066357, 146.02729738])

In [50]:
ss.Sales = results_cat
ss.to_csv("submission_cat_Yeqing_Xia.csv", index=False)

In [51]:
ss.Sales = 0.5*results_cat + 0.2*results_xgb + 0.3*results_cat
ss.to_csv("submission_stacked_Yeqing_Xia.csv", index=False)

In [52]:
ss

ID       Sales
0         883  120.949105
1         884  109.236224
2         885  103.644612
3         886  105.261839
4         887   56.210263
5         888   84.294407
6         889   76.943572
7         890   74.441534
8         891   69.431447
9         892   70.419817
10        893   75.041029
11        894   56.154256
12        895   68.321729
13        896  143.195674
14        897  113.958059
15        898  104.243399
16        899  103.633689
17        900  103.033227
18        901   57.583116
19        902   62.941776
20        903   76.134264
21        904   70.561520
22        905   73.939262
23        906   71.980311
24        907   76.490456
25        908   55.117008
26        909   72.752718
27        910  151.808669
28        911  117.160404
29        912  104.047020
30        913   98.926253
31        914   94.116052
32        915   57.860933
33        916   73.317486
34        917   75.815488
35        918   72.472860
36        919   69.088859
37        920   70.833182
38        921   70.891258
39        922   56.567750
40        923   79.197035
41        924  132.739324
42        925  110.159392
43        926  100.464441
44        927   96.231741
45        928   96.466691
46        929   55.033387
47        930   78.640974
48        931   76.961198
49        932   74.642654
50        933   76.761398
51        934   75.729259
52        935   75.744766
53        936   54.041283
54        937   74.012769
55        938  142.979745
56        939  117.961472
57        940  105.895913
58        941  100.362417
59        942   96.126783
60       1825  161.528074
61       1826  146.852929
62       1827   95.813713
63       1828  150.844941
64       1829   78.414833
65       1830  116.141435
66       1831  107.641245
67       1832  104.747046
68       1833  100.044554
69       1834  100.677796
70       1835  111.656594
71       1836   74.755871
72       1837  101.628222
73       1838  181.922563
74       1839  152.225994
75       1840  144.632298
76       1841  147.837238
77       1842  149.319083
78       1843   75.291851
79       1844   95.667074
80       1845  107.361100
81       1846   99.859088
82       1847  103.374852
83       1848  106.370015
84       1849  113.421408
85       1850   72.885594
86       1851  113.812213
87       1852  198.062832
88       1853  158.714167
89       1854  142.380883
90       1855  140.142079
91       1856  140.662319
92       1857   81.488940
93       1858  113.443168
94       1859  111.875052
95       1860  107.433992
96       1861  104.041620
97       1862  109.651426
98       1863  117.531334
99       1864   89.530411
...       ...         ...
35900  547106  109.795068
35901  547107  101.460571
35902  547108  107.663439
35903  547109  110.238513
35904  547110  119.893341
35905  547111   96.555821
35906  547112  110.691232
35907  547113  180.450001
35908  547114  142.957933
35909  547115  132.837680
35910  547116  132.169403
35911  547117  132.437637
35912  547118  105.650126
35913  547119  110.453627
35914  547120  116.498182
35915  547121  107.799881
35916  547122  103.523019
35917  547123  106.496675
35918  547124  111.017966
35919  547125  101.775280
35920  547126  117.373152
35921  547127  157.771497
35922  547128  132.231143
35923  547129  125.213590
35924  547130  125.551161
35925  547131  129.861964
35926  547132   99.928548
35927  547133  101.418650
35928  547134  111.163080
35929  547135  108.363628
35930  547136  113.106943
35931  547137  113.369537
35932  547138  122.618736
35933  547139  108.011114
35934  547140  115.458498
35935  547141  173.224406
35936  547142  145.998274
35937  547143  135.042463
35938  547144  133.524347
35939  547145  132.590627
35940  548028  152.270052
35941  548029  143.192825
35942  548030  123.827489
35943  548031  162.433848
35944  548032  122.874915
35945  548033  123.680335
35946  548034  106.842147
35947  548035  102.074517
35948  548036   98.814853
35949  548037   97.656560
35950  548038  114.802194
35951  548039  106.742585
35952  